### PR TITLE
[BIOMAGE-1236] - Download Processed Seurat Object

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1236

#### Link to staging deployment URL 

https://ui-agi-download-seurat.scp-staging.biomage.net/

#### Links to any Pull Requests related to this

UI - https://github.com/biomage-ltd/ui/pull/416
API - https://github.com/biomage-ltd/api/pull/194

#### Anything else the reviewers should know about the changes here

- Download is done by preparing a signed download link that allows direct download from AWS S3.

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [x] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/components/data-management/ProjectDetails.jsx [hunk 1]
```diff
@@ -577,22 +622,30 @@ const ProjectDetails = ({ width, height }) => {
...
+            Processed Seurat object (.rds)
+          </a>
         </Tooltip>
       </Menu.Item>
-      <Menu.Item disabled>
-        <Tooltip title='Coming soon!' placement='left'>
-          Data Processing settings
-        </Tooltip>
+      <Menu.Item
+        disabled
```
Comment (kafkasl) [inline src/components/data-management/ProjectDetails.jsx]: add coming soon tooltip here too

File: src/components/data-management/ProjectDetails.jsx [hunk 2]
```diff
@@ -577,22 +622,30 @@ const ProjectDetails = ({ width, height }) => {
...
       </Menu.Item>
-      <Menu.Item disabled>
-        <Tooltip title='Coming soon!' placement='left'>
-          {/* <Tooltip title='Data Processing filters have been applied' placement='left'> */}
-          Processed Seurat object (.rds)
+      <Menu.Item
+        key='download-processed-seurat'
+        disabled={!pipelineHasRun}
+      >
+        <Tooltip title='Data Processing filters have been applied' placement='left'>
```
Comment (kafkasl) [inline src/components/data-management/ProjectDetails.jsx]: I think this tooltip is not very clear. I would probably write something like "Seurat object produced by applying data processing filters" (not very happy with that  phrasing either. Do you have any suggestion @vickymorrison ?

File: src/pages/data-management/index.jsx [hunk 3]
```diff
@@ -54,6 +55,13 @@ const DataManagementPage = ({ route }) => {
   }, []);
 
   useEffect(() => {
+    if (activeProject?.experiments.length > 0) {
+      // Right now we have one experiment per project, so we can just load the experiment
+      // This has to be changed when we have more than one experiment
+      const activeExperimentId = activeProject.experiments[0];
+      dispatch(loadBackendStatus(activeExperimentId));
```
Comment (kafkasl) [inline src/pages/data-management/index.jsx]: why is this needed?

File: src/redux/reducers/experiments/experimentsDataDownloaded.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+const experimentsDataDownload = (state, action) => {
```
Comment (kafkasl) [inline src/redux/reducers/experiments/experimentsDataDownloaded.js]: shouldn't this reducer use the new method that @cosa65 introduced to avoid the state unpacking?

**Review by kafkasl — CHANGES_REQUESTED** [review]

File: src/components/data-management/ProjectDetails.jsx (hunk 1 above)
Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx]: added

File: src/components/data-management/ProjectDetails.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx]: "Data with filters applied"?

File: src/pages/data-management/index.jsx (hunk 3 above)
Comment (aerlaut) [reply src/pages/data-management/index.jsx]: To check whether pipeline has been run for this experiment. To do this, the status of the `pipeline` run is checked. Backend status was not loaded for this module.

File: src/redux/reducers/experiments/experimentsDataDownloaded.js (hunk 4 above)
Comment (aerlaut) [reply src/redux/reducers/experiments/experimentsDataDownloaded.js]: Whoops, these reducers should be removed

File: src/components/data-management/ProjectDetails.jsx (hunk 2 above)
Comment (vickymorrison) [reply src/components/data-management/ProjectDetails.jsx]: How about "With Data Processing filters and settings applied"?

File: src/components/data-management/ProjectDetails.jsx [hunk 5]
```diff
@@ -104,6 +115,47 @@ const ProjectDetails = ({ width, height }) => {
     }
   }, [samples, activeProject]);
 
+  const downloadData = async (type) => {
+    const experimentId = activeProject?.experiments[0];
+
+    try {
+      if (!environment) throw new Error('Environment is not set');
```
Comment (kafkasl) [inline src/components/data-management/ProjectDetails.jsx]: why is this needed? the variable is not being used in the method

File: src/components/data-management/ProjectDetails.jsx [hunk 6]
```diff
@@ -64,15 +71,19 @@ const ProjectDetails = ({ width, height }) => {
...
   const activeProject = useSelector((state) => state.projects[activeProjectUuid]) || false;
+  const { backendStatus } = useSelector((state) => state.experimentSettings);
+
   const [tableData, setTableData] = useState([]);
   const [tableColumns, setTableColumns] = useState([]);
   const [sortedSpeciesData, setSortedSpeciesData] = useState([]);
   const [sampleNames, setSampleNames] = useState(new Set());
   const [canLaunchAnalysis, setCanLaunchAnalysis] = useState(false);
   const [analysisModalVisible, setAnalysisModalVisible] = useState(false);
+  const [pipelineHasRun, setPipelineHasRun] = useState(true);
```
Comment (kafkasl) [inline src/components/data-management/ProjectDetails.jsx]: why is the default `true`? shouldn't we assume that the pipeline has not run?

File: src/components/data-management/ProjectDetails.jsx [hunk 7]
```diff
@@ -104,6 +115,47 @@ const ProjectDetails = ({ width, height }) => {
...
+    if (
+      !loading
+      && !error
+      && status?.pipeline
+      && status.pipeline?.status === pipelineStatusValues.SUCCEEDED
+    ) {
+      setPipelineHasRun(true);
+    }
+
+    setPipelineHasRun(false);
```
Comment (kafkasl) [inline src/components/data-management/ProjectDetails.jsx]: this is being defined on a project-level. This will break once there is more than a single-experiment per project. We are already having issues where different projects interact among them so I'd be very careful to limit the visibility of state to the relevant component.

File: src/components/data-management/ProjectDetails.jsx [hunk 8]
```diff
@@ -104,6 +115,47 @@ const ProjectDetails = ({ width, height }) => {
     }
   }, [samples, activeProject]);
 
+  const downloadData = async (type) => {
+    const experimentId = activeProject?.experiments[0];
```
Comment (kafkasl) [inline src/components/data-management/ProjectDetails.jsx]: I think the logic of choosing the experiment to download should not be inside this function it just does too many things inducing errors. This function should receive a type and the experimentID (or whatever it's needed) and then just download. The logic about which is the active experiment shouldn't be here. I suspect that this kind of accesses are the ones causing projects to interfere with one another. You click one experiment but the activeProject is not the correct one. If every project passes explicitly it's own ID this will not happen.

File: src/pages/data-management/index.jsx [hunk 9]
```diff
@@ -54,6 +55,13 @@ const DataManagementPage = ({ route }) => {
   }, []);
 
   useEffect(() => {
+    if (activeProject?.experiments.length > 0) {
```
Comment (kafkasl) [inline src/pages/data-management/index.jsx]: what happens if experimens is undefined? can it happen?

**Review by kafkasl — CHANGES_REQUESTED** [review]

File: src/pages/data-management/index.jsx (hunk 9 above)
Comment (aerlaut) [reply src/pages/data-management/index.jsx]: No, `experiments` default to an empty array

File: src/components/data-management/ProjectDetails.jsx (hunk 6 above)
Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx]: oh yea, this is supposed to be `false`. Changed accordingly,

File: src/pages/data-management/index.jsx (hunk 3 above)
Comment (kafkasl) [reply src/pages/data-management/index.jsx]: actually this is not good because of at least 2 reasons.
1. this isn't awaited so you can be effectively acting on the backend status of the previous experiment
2. this is called every time you change an experiment which is a waste. 

What I suggest (but might be out of the scope of the PR) is to store the backend statuses indexed by experimentID so you just load them once and avoid the potential issue of using the wrong status. If you don't do that then at least `await` the backend status load if possible

File: src/components/data-management/ProjectDetails.jsx (hunk 5 above)
Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx:122]: Oh yea, it's not used anymore. I'll remove it,

File: src/components/data-management/ProjectDetails.jsx (hunk 8 above)
Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx]: Refactored to another function. It now takes experimentId as a parameter

**Comment by vickymorrison** [comment]
Suggestion (but not critical): when the rds object is downloaded it appears in my downloads as "r.rds". It would be much better if the file is named with the project name and "processed" e.g. 'ProjectID_processed_object' or similar. The user might download both the raw and processed objects from multiple projects and it would be much better UX if it is clear which is which.

**Comment by aerlaut** [comment]
> Suggestion (but not critical): when the rds object is downloaded it appears in my downloads as "r.rds". It would be much better if the file is named with the project name and "processed" e.g. 'ProjectID_processed_object' or similar. The user might download both the raw and processed objects from multiple projects and it would be much better UX if it is clear which is which.

Sure. The projectId is quite long, so I'll only use the first part of the projectId (e.g. if projectId = AA-BB-CC-DD-EE, I'll use AA for the file name).

Added in https://github.com/biomage-ltd/api/pull/194/commits/fe3570d351163713de2e94e1f4cb1e13f7cb78f5

File: src/components/data-management/ProjectDetails.jsx (hunk 7 above)
Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx]: Refactored in 08d66f3

File: src/components/data-management/ProjectDetails.jsx [hunk 10]
```diff
@@ -575,25 +582,50 @@ const ProjectDetails = ({ width, height }) => {
...
+          title={
+            activeProject?.experiments?.length > 0
+            && pipelineHasRun(activeProject.experiments[0])
+              ? 'With Data Processing filters and settings applied'
+              : 'Launch analysis to process data'
+          }
+          placement='left'
+          onClick={() => {
+            // Change if we have more than one experiment per project
+            const experimentId = activeProject.experiments[0];
```
Comment (kafkasl) [inline src/components/data-management/ProjectDetails.jsx]: shouldn't this be `activeProject?.experiments[0]`

**Review by kafkasl — APPROVED** [review]

File: src/components/data-management/ProjectDetails.jsx (hunk 10 above)
Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx]: Existence of experiments[0] is checked on the first clause, so it should be available.

Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx]: The menu item is now disabled if there is no experimentId, so the click does not work if there is no experimentId.

**Review by StefanBabukov — APPROVED** [review]

File: src/components/data-management/ProjectDetails.jsx (hunk 10 above)
Comment (aerlaut) [reply src/components/data-management/ProjectDetails.jsx]: change accordingly
